### PR TITLE
SDKS-1760_Dynamic_sdk_configuration

### PR DESCRIPTION
**Pull request by george-bafaloukas-forgerock**
# JIRA Ticket

Please, link jira ticket here.

# Description

Briefly describe the change and any information that would help speedup the review and testing process.

# Definition of Done Checklist:

- [ ] Acceptance criteria is met.
- [ ] All tasks listed in the user story have been completed.
- [ ] Coded to standards.
- [ ] Ensure backward compatibility.
- [ ] API reference docs is updated.
- [ ] Unit tests are written.
- [ ] Integration tests are written.
- [ ] e2e tests are written.
- [ ] Functional spec is written/updated.
- [ ] Example code snippets have been added.
- [ ] Change log updated.
- [ ] Documentation story is created and tracked.
- [ ] Tech debts and remaining tasks are tracked in separated ticket(s).

### CODE REVIEW

**Review by jeyanthanperiyasamy — COMMENTED** [review]
@george-bafaloukas-forgerock the one scenario we discussed is,  when they call FRAuth.start multiple times with same domain name with different configuration . we may need to revoke the token.

File: FRAuth/FRAuth/FRAuth.swift [hunk 1]
```diff
@@ -85,6 +96,22 @@ public final class FRAuth: NSObject {
     /// - Parameter config: Dictionary object of configuration
     /// - Throws: ConfigError
     static func initPrivate(config: [String: Any]) throws {
+        //Check if there is an existing session and destroy it
+        if let currentUser = FRUser.currentUser {
+            if let _ = FRAuth.shared?.oAuth2Client {
+                let semaphore = DispatchSemaphore(value: 1)
+                //Revoke the Access Token
+                currentUser.revokeAccessToken { _, _ in
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/FRAuth.swift:104]: also check is there any memory leak in this line

File: FRAuth/FRAuth/Manager/KeychainManager.swift [hunk 2]
```diff
@@ -226,6 +226,31 @@ struct KeychainManager {
...
+    
+    /// Stores SSOToken into designated Keychain Service, or removes SSOToken when nil
+    /// - Parameter ssoToken: Token object
+    /// - Returns: Boolean result of operation
+    @discardableResult func setFROptions(frOptionsData: Data?) -> Bool {
+        if let frOptions = frOptionsData {
+            return self.privateStore.set(frOptions, key: "FROptions")
+        }
+        else {
+            return self.privateStore.delete("FROptions")
```
Comment (vahancouver) [inline FRAuth/FRAuth/Manager/KeychainManager.swift:250]: I think we should start extracting string literals/constants into let constants. Here "FROptions" is used 3 times.

File: FRAuth/FRAuth/FRAuth.swift [hunk 3]
```diff
@@ -86,6 +97,13 @@ public final class FRAuth: NSObject {
...
         
+        // Check if there is an existing session and destroy it, do a clean up as this would only happen at this point
+        // if the SDK has been previously initialized and a succesfull authentication has happened. Given that the SDK gets
+        // re-initiallised, possibly with a different configuration we need to do some clean up.
+        if let _ = FRUser.currentUser {
+            FRAuth.cleanUp()
+        }
+        
         // Validate server config
         guard let server = config["forgerock_url"] as? String,
```
Comment (vahancouver) [inline FRAuth/FRAuth/FRAuth.swift:108]: Inside this method we use a lot of hard coded config property names. Now that we have FROptions.CodingKeys enum, I think we should use that instead: **FROptions.CodingKeys.url** in this case.

**Review by vahancouver — COMMENTED** [review]
Overall looks great. 

Some minor comments.

**Review by spetrov — APPROVED** [review]

File: FRAuth/FRAuth/Config/FROptions.swift [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+//  FROptions.swift
+//  FRAuth
+//
+//  Copyright (c) 2022 ForgeRock. All rights reserved.
+//
+//  This software may be modified and distributed under the terms
+//  of the MIT license. See the LICENSE file for details.
+//
+
+import Foundation
```
Comment (rodrigoareis) [inline FRAuth/FRAuth/Config/FROptions.swift:11]: Do we need API doc for this class?

File: FRAuth/FRAuth/FRAuth.swift [hunk 5]
```diff
@@ -39,6 +39,8 @@ public final class FRAuth: NSObject {
     /// Shared instance of FRAuth
     @objc
     public static var shared: FRAuth? = nil
+
```
Comment (rodrigoareis) [inline FRAuth/FRAuth/FRAuth.swift:42]: Update copyright

File: FRAuth/FRAuth/Manager/KeychainManager.swift [hunk 6]
```diff
@@ -38,6 +38,8 @@ struct KeychainManager {
     let defaultService: String = "com.forgerock.ios.keychainservice"
     /// String constant for default bundle identifier
     let defaultBundleIdentifier: String = "com.forgerock.ios.sdk"
+    /// String constant for FROptions storage key
```
Comment (rodrigoareis) [inline FRAuth/FRAuth/Manager/KeychainManager.swift:41]: Update copyright

File: FRAuth/FRAuth/Manager/KeychainManager.swift [hunk 7]
```diff
@@ -226,6 +228,31 @@ struct KeychainManager {
         return nil
     }
     
+    //  MARK: - FROptions
+    
+    /// Returns current session's Token object that represents SSO Token
```
Comment (rodrigoareis) [inline FRAuth/FRAuth/Manager/KeychainManager.swift:233]: Returns session's token?

File: FRAuth/FRAuthTests/FRAuthSwiftTests/FRAuth/FRSession/FRSessionTests.swift [hunk 8]
```diff
@@ -464,7 +464,7 @@ class FRSessionTests: FRAuthBaseTest {
         FRAuth.configPlistFileName = "FRAuthConfigNoOAuth2"
         self.config = thisConfig
         self.startSDK()
-        
+        super.setUp()
```
Comment (rodrigoareis) [inline FRAuth/FRAuthTests/FRAuthSwiftTests/FRAuth/FRSession/FRSessionTests.swift:467]: Update copyright

File: FRProximity/FRProximityTests/E2ETests/DeviceCollectorNodeTests.swift [hunk 9]
```diff
@@ -25,6 +25,7 @@ class DeviceCollectorNodeTests: FRAuthBaseTest {
         do {
             // Init SDK
             try FRAuth.start()
+            super.setUp()
```
Comment (rodrigoareis) [inline FRProximity/FRProximityTests/E2ETests/DeviceCollectorNodeTests.swift:28]: Update copyright

**Review by rodrigoareis — COMMENTED** [review]
Looks good to me. Left some comments.

File: FRAuth/FRAuth/Config/FROptions.swift [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    public var enableCookie: Bool
+    public var cookieName: String
+    public var timeout: String
+    
+    public var authenticateEndpoint: String?
+    public var authorizeEndpoint: String?
+    public var tokenEndpoint: String?
+    public var revokeEndpoint: String?
+    public var userinfoEndpoint: String?
+    public var sessionEndpoint: String?
```
[the remark this answers is not in the document]
Comment (george-bafaloukas-forgerock) [reply FRAuth/FRAuth/Config/FROptions.swift:26]: Let me have a look. I didn't add/remove anything from the existing FRAuth.start implementation. So if it wasn't there before, I haven't added it as well

File: FRAuth/FRAuth/Manager/KeychainManager.swift (hunk 7 above)
Comment (george-bafaloukas-forgerock) [reply FRAuth/FRAuth/Manager/KeychainManager.swift:233]: good catch, there is one more comment that needs fixing

File: FRAuth/FRAuth/Config/FROptions.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (george-bafaloukas-forgerock) [reply FRAuth/FRAuth/Config/FROptions.swift:26]: @spetrov the "set(endSessionPath:)" method was not used in our SDK. Searching the repo, does not surface any results other than tests. It must have been missed, unless it was omitted intentionally in the past for reasons unknown to me. I will add it in the FRAuth.start and the FROptions for completion.

[the remark this answers is not in the document]
Comment (george-bafaloukas-forgerock) [reply FRAuth/FRAuth/Config/FROptions.swift:28]: Done, please re - review

File: FRAuth/FRAuth/FRAuth.swift (hunk 5 above)
Comment (george-bafaloukas-forgerock) [reply FRAuth/FRAuth/FRAuth.swift:42]: done

File: FRAuth/FRAuth/Manager/KeychainManager.swift (hunk 7 above)
Comment (george-bafaloukas-forgerock) [reply FRAuth/FRAuth/Manager/KeychainManager.swift]: done

File: FRAuth/FRAuthTests/FRAuthSwiftTests/FRAuth/FRSession/FRSessionTests.swift (hunk 8 above)
Comment (george-bafaloukas-forgerock) [reply FRAuth/FRAuthTests/FRAuthSwiftTests/FRAuth/FRSession/FRSessionTests.swift:467]: done

File: FRAuth/FRAuth/Config/FROptions.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (spetrov) [reply FRAuth/FRAuth/Config/FROptions.swift:28]: 👍🏻 Thanks! Confirmed that it was added and works as expected! I was able to trigger a flow that uses this endpoint...

**Review by vahancouver — APPROVED** [review]
Changes look good.

Are we also updating the SDK docs to talk about this option of initializing the SDK too?

**Comment by george-bafaloukas-forgerock** [comment]
> Changes look good.
> 
> Are we also updating the SDK docs to talk about this option of initializing the SDK too?

Yeah I think we need a separate ticket for that

File: FRAuth/FRAuthTests/FRAuthSwiftTests/FRAuth/RequestInterceptor/FRRequestInterceptorTests.swift [hunk 11]
```diff
@@ -19,107 +19,29 @@ class FRRequestInterceptorTests: FRAuthBaseTest {
     override func setUp() {
         self.configFileName = "Config"
         super.setUp()
+        self.callLogin()
```
Comment (rodrigoareis) [inline FRAuth/FRAuthTests/FRAuthSwiftTests/FRAuth/RequestInterceptor/FRRequestInterceptorTests.swift:22]: Update copyright

**Review by rodrigoareis — COMMENTED** [review]
All changes looks good. Just one class updated in latest commit needs copyright update.

**Review by rodrigoareis — APPROVED** [review]
Looks good to me

**Pull request merged**